lib/Frontend/CompilerInvocation.cpp

```diff
@@ -215,11 +201,34 @@ class ArgsToFrontendInputsConverter {
     return true;
   }
 
-  llvm::StringSet<> getPrimaries() const {
-    llvm::StringSet<> primaryFiles;
+  bool readPrimaryFiles() {
     for (const Arg *A : Args.filtered(options::OPT_primary_file))
-      primaryFiles.insert(A->getValue());
-    return primaryFiles;
+      PrimaryFiles.insert(A->getValue());
+    if (forAllFilesInFilelist(
+            PrimaryFilelistPathArg,
+            [&](StringRef file) -> void { PrimaryFiles.insert(file); }))
+      return true;
+    return false;
+  }
+
+  void createInputFiles() {
+    for (auto &file : Files) {
+      bool isPrimary = PrimaryFiles.count(file) > 0;
+      Inputs.addInput(InputFile(file, isPrimary));
+      if (isPrimary)
+        PrimaryFiles.erase(file);
```

Comment:
Hm. I don't like that this destroys `PrimaryFiles`. It's a convenient way to count primaries that don't correspond to the filelist, but it makes the whole data structure stateful in a way that it wouldn't be if the set were a local variable. Maybe `checkForMissingPrimaryFiles` should just be folded into this function.

Reply:
BuffersToKeepAlive, Files, and PrimaryFiles are all stateful in this data structure. I could, by passing references, make them all local variables to convert, or just make some of them local. PrimaryFiles is the only one that both grows and shrinks, as you notice. Is that enough of a distinction to single it out for local-variable status? Happy to do so if you like, just felt odd to me to single that one out.

Reply:
BuffersToKeepAlive doesn't bother me because it's an implementation detail. Files is a little funnier, but it still only really has one transition point: it starts off empty, and then once the list of files is calculated it's set correctly forever after. PrimaryFiles having more states makes me uncomfortable, particularly when you can't tell what state it's in. (I guess you could never tell if it was actually absent or just not-yet-computed, but it gets set pretty early on.)

Reply:
OK, I've made it a local and clarified the method names a bit.